### ISSUE TITLE
Support for multiple back-ends and redis client libraries?

### ISSUE THREAD

**Issue by b3ntly** [b3ntly/bucket#2]
I received a suggestion on Reddit that this library should support multiple Redis libraries by abstracting various methods like .Get and .Set to a common interface. I definitely agree with this, in fact (though it predates my initial commit) the original implementation used an interface called Storage to do exactly that, though it was intended to abstract the database system itself (to support memcached and Redis) and not just Redis client libraries.

The API looked like this:

```golang
storage, err := NewStorage(options interface{})
bucket, err := NewBucket(name string, capacity int, storage *Storage)
```

However I pulled back for the initial release for a few reasons:

* The library currently had 0 users and I wanted to keep the surface area small so I could focus on core features

* The library currently uses the lock-less approach of lua scripting to maintain thread safety thus any other redis library would preferably support Lua scripting and any other database system would likely need a fair bit of custom logic to maintain feature parity and quality.

* I think theirs some turmoil in Go where newcomers struggle to choose the "best" third-party library for a given task. I think many of the redis libraries do have feature parity but in my opinion Go-redis (a.k.a. gopkg.in/redis.v6) has been the best maintained and implemented redis library for Golang for the past few years.

So in short yes I will be implementing an interface that will support this but not in the next few weeks for the reasons stated above. First priority of the "Storage" interface will be supporting an in-memory version of the token bucket, then support for memcached (because that would allow Google App Engine users to implement a token bucket with GAE's mostly free managed memcached service which is cool), and finally for Redis client libraries.

As always thanks for the feedback and let me know if their are any more questions on the topic.

**Comment by b3ntly** [comment]
Quick follow up, I added the proposed interface (found in storage.go) and moved over all logic for Redis and the newly added in-memory store to use it. It should be fairly trivial to add many more storage adapters using redis.go and memory.go as templates. If anyone actually does this please don't forget to add your provider to the table based testing so all applicable tests can be run against a new provider.
